src/vs/workbench/contrib/terminalContrib/suggest/browser/terminalCompletionService.ts

```diff
@@ -57,31 +63,46 @@ export class TerminalCompletionService extends Disposable implements ITerminalCo
 		});
 	}
 
-	async provideCompletions(promptValue: string, cursorPosition: number, shellType: TerminalShellType): Promise<ISimpleCompletion[] | undefined> {
+	async provideCompletions(promptValue: string, cursorPosition: number, shellType: TerminalShellType, token: CancellationToken, triggeredProviders?: ITerminalCompletionProvider[]): Promise<ISimpleCompletion[] | undefined> {
 		const completionItems: ISimpleCompletion[] = [];
 
 		if (!this._providers || !this._providers.values) {
 			return undefined;
 		}
 
-		// TODO: Use Promise.all so all providers are called in parallel
-		for (const providerMap of this._providers.values()) {
-			for (const [extensionId, provider] of providerMap) {
+		const collectCompletions = async (providers: ITerminalCompletionProvider[]) => {
+			await Promise.all(providers.map(async provider => {
 				if (provider.shellTypes && !provider.shellTypes.includes(shellType)) {
-					continue;
+					return;
 				}
-				const completions = await provider.provideCompletions(promptValue, cursorPosition);
+				const completions = await provider.provideCompletions(promptValue, cursorPosition, token);
 				const devModeEnabled = this._configurationService.getValue(TerminalSettingId.DevMode);
 				if (completions) {
 					for (const completion of completions) {
-						if (devModeEnabled && !completion.detail?.includes(extensionId)) {
-							completion.detail = `(${extensionId}) ${completion.detail ?? ''}`;
+						if (devModeEnabled && !completion.detail?.includes(provider.id)) {
+							completion.detail = `(${provider.id}) ${completion.detail ?? ''}`;
 						}
 						completionItems.push(completion);
```

Comment:
It would be cleaner if we providers returned promises of the completions, then they get merged after `Promise.all`